### ISSUE TITLE
Identifying power consumption of hardware on the PCB

### ISSUE THREAD

**Issue by bradanlane** [fdufnews/SMART-Response-XE-schematics#1]
I've been working with several different "libraries" for the SMART Response XE.

All of the repositories I have been able to test have the same power issue.

When I test the power consumption of the original code on a production device, the steady state is about 1.5mA (with content visible on the LCD) and the sleep state is about 200uA.

All of the repositories I have tested have a steady state is about 7-9mA and the sleep state is 500-600uA.

Have you found any documentation which may help identify pins which are enabling or providing power drain. It's not clear to me if the issue is the LCD or a pin controlling some other hardware on the PCB.

My test code does not initialize any of the libraries and it only changes PORTn and DDRn values for pins. I can get down to 4-5mA but no lower.

**Comment by hotkeymuc** [comment]
Without having tested power drain myself (using the device tethered) it may be down to putting the RF unit into sleep mode.
Have you tried rfSleep()? My local version of "RadioFunctions.h" contains it. Here is a hard copy of it:

```
void rfSleep() {
  
  // Store old state
  ant_div = ANT_DIV;
  ANT_DIV = 0 | 3;  // Disable antenna diversity (0x03)
  delay(1);
  
  TRX_STATE = (TRX_STATE & 0xE0) | TRX_OFF;  // Set to TRX_OFF state
  while ((TRX_STATUS & 0x1F) != TRX_OFF) {
    delay(1);
  }

  // Sleep
  TRXPR |= (1 << 1);  // Set SLPTR bit, going from TRX_OFF to SLEEP, turning the Transceiver completely OFF
  delay(1);
}
```
Hope it helps!

**Comment by bradanlane** [comment]
I have augmented some existing RF code so I have control over its state.

The RF transceiver adds another 9-10mA on top of the steady state. Commercial code goes up to 12-13mA and open source libraries increase to 18-19mA.

**Comment by hotkeymuc** [comment]
So I assume you are already calling the "SRXESleep()" function which puts LCD, ADC and CPU into SLEEP_MODE_PWR_DOWN mode.
The last thing would be some kind of quiescent drain over one of the communication lines connected to an outside device.
But that's all I could think of.

**Comment by bradanlane** [comment]
I have used a few different SRXE libraries and have made some changes to them to achieve minor improvements in power, but only 1-2mA.

There are two issues (and they may be related).

When the device is up an running, the commercial code is only using 2mA and that is while the LCD has content visible. The various repository code is 6-8mA as a minimum.

So there is clearly something the commercial code is doing to dramatically reduce power.

Even when the RF is active, the commercial code is about 6mA more efficient.

Then, in deep sleep, the SRXE library is 500-600uA which the commercial code is 200uA.

My assumption is there is hardware on the PCB which is controllable via a pin which the commercial code is properly setting to either turn off an unused circuit or prevent a circuit from draining power. I'm just not finding it.

**Comment by fdufnews** [comment]
I just made some measurement on my SRXE running the [basic interpreter](https://github.com/fdufnews/SMART_Response_BASIC) I have adapted and found 5.9mA in idle mode and 127µA in sleep mode.
The [low level library](https://github.com/fdufnews/SMART-Response-XE-Low_level) on which the basic is relying is based on [bitbank2 library](https://github.com/bitbank2/SmartResponseXE) with minor modification.
The idle mode is tightly depending on the application. For the basic interpreter for example the processor is always running the low level keyboard scanning. An application that would like to be energy efficient would probably use a timer to wake the processor every 20ms or 50ms, make a keyboard scan and return in sleep mode if no action was detected.
I am not using the RF part of the SRXE.

**Comment by bradanlane** [comment]
I will have to look closely at basic code. I can get to 5mA but sleep is 500mA.

I don't know how the commercial code is at or below 2mA when running.

**Comment by fdufnews** [comment]
The magic is not in the Basic code but in the low level library.
Look at the SRXESleep() and SRXEPowerDown() [here](https://github.com/fdufnews/SMART-Response-XE-Low_level/blob/80f448ce9b5e37ba63baa57b1414111d55b3a0dd/avr/libraries/SmartResponseXE/SmartResponseXE.cpp#L326)  and [here](https://github.com/fdufnews/SMART-Response-XE-Low_level/blob/80f448ce9b5e37ba63baa57b1414111d55b3a0dd/avr/libraries/SmartResponseXE/SmartResponseXE.cpp#L506)

**Comment by bradanlane** [comment]
I will have to consider this a difference in our measurement methodology.

I built the "lowlevel" library and run `hello_world` code. I am recording 7mA while running and 400uA when in deep sleep.

I still can not figure out how the original commercial code is below 2mA when active. That is the real mystery. It's putting something to sleep while preserving the display. If it's "working on something" then it's at 6mA so the same as your low level code.

**Comment by fdufnews** [comment]
For the measurement I made last week I was using a dollar store multimeter and a phone charger as power supply.
I have made more investigation with a more accurate multimeter and adjustable power supply. The current consumption depends on the voltage of the battery. But the results are not that far from what I measured the first time. 
 I have made measures with the hello_world application
```
Battery       idle            sleep
 3V           7.6mA           280uA
 4V           8mA             200uA
 5V           6.6mA           160uA
 6V           5.3mA           150uA
```

**Comment by fdufnews** [comment]
Made modification to https://github.com/fdufnews/SMART-Response-XE-Low_level to reduced sleeping current down to 60uA @5V by programming the prescaler to its maximum value before entering sleep.
Concerning idle mode I think it will be more difficult. Still investigating.

**Comment by fdufnews** [comment]
I probably made some mistake while testing the previous modification.
I can't reproduce the 60uA. The frequency change does nothing. I am asking myself if the processor is really in deep sleep.
I have played with the PRRx registers and found no improvement to the current during deep sleep.
The quiescent current of the DC/DC converter is supposed to be between 15 and 30uA. I can't find which part uses all that current.
I have deleted the modification in the library and I am still investigating.
